net.rocketpool/image/src/app.ts

```diff
@@ -1,163 +1,262 @@
+      //   return false;
+      // }
+      // ```
+      // But in our case, we cannot act so naively -- there can be multiple events in one block:
+      // ```
+      // ┌─────────────┐
+      // │ E1 E2 E3 E4 │
+      // └─────────────┘
+      //  Block №1
+      // ```
+      //
+      // This gives us an interesting edge case if the batch size goes through the middle of the block:
+      // ```
+      //  Batch №1  ┆ Batch №2
+      // ┌───────── ┆ ───┐  ┌─────────────┐
+      // │ E1 E2 E3 ┆ E4 │  │ E5 E6 E7 E8 │
+      // └───────── ┆ ───┘  └─────────────┘
+      //  Block №1  ┆        Block №2
+      // ```
+      //
+      // Since, our next read iteration relies on a range of blocks, this situation creates ambiguity for us:
+      // should we consider this block as processed or not?
+      //
+      // For this reason, we read additional events to consider this block as fully processed:
+      // ```
+      //  Batch №1         ┆  Batch №2
+      // ┌─────────   ───┐ ┆ ┌─────────────┐
+      // │ E1 E2 E3 + E4 │ ┆ │ E5 E6 E7 E8 │
+      // └─────────   ───┘ ┆ └─────────────┘
+      //  Block №1         ┆  Block №2
+      // ```
+      const isEventFromSameBlock = lastOutputBlockNumber === eventRow.blockNumber;
+
+      if (hasBatchAlreadyCollected && !isEventFromSameBlock) {
+        return false;
+      }
+
+      outputRowsCount++;
+      lastOutputBlockNumber = eventRow.blockNumber;
+
+      if (!hasBatchAlreadyCollected) {
+        hasBatchAlreadyCollected = outputRowsCount >= odfBatchSize;
+      }
+
+      process.stdout.write(JSON.stringify(eventRow));
+      process.stdout.write("\n");
+
+      return true;
+    },
+    blockBatchSize,
   );
 
+  const etag = allBlockRangeHasProcessed ? endBlock : lastOutputBlockNumber;
+  const hasMore = !allBlockRangeHasProcessed;
+
+  return {
+    etag,
+    hasMore
+  };
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+async function readEventsFromETH(
+  rethContract: RethAbi,
+  startBlock: number,
+  endBlock: number,
+  onEventCallback: OnEventCallback,
+  blockBatchSize: number
+): Promise<AllBlockRangeHasProcessed> {
+  const blocksDelta = endBlock - startBlock;
+
+  console.log(`Considering events in block range [${startBlock}, ${endBlock}] (${blocksDelta} blocks)`);
+
   let fromBlock = startBlock;
   let toBlock = fromBlock;
 
   do {
     fromBlock = toBlock + 1;
-    toBlock = Math.min(endBlock, fromBlock + stride);
+    toBlock = Math.min(endBlock, fromBlock + blockBatchSize);
+
+    console.log(`Syncing block range [${fromBlock}, ${toBlock}]`);
 
-    console.log("Syncing block range [%s, %s]", fromBlock, toBlock);
-    await readEventsFromETHRangeWithFallback(fromBlock, toBlock, clb);
+    const stopProcessing = !await readEventsFromETHRangeWithFallback(
+      rethContract, fromBlock, toBlock, onEventCallback
+    );
+
+    if (stopProcessing) {
+      return false;
+    }
   } while (toBlock != endBlock);
```

Comment:
to keep best practices and never compare `NaN`please change to !==

Reply:
A good catch! Fixed